### PR TITLE
Porting ConfigParser to use std::string_view wherever possible

### PR DESCRIPTION
**Pull request by Durganshu**
## Main changes of this PR

The ConfigParser bridges the call-back-based C-library libxml2 to the C++ XMLTag and XMLAttribute of preCICE. This PR ports ConfigParser to use std::string_view instead of std::string where the components don't own data. This leads to less casting, fewer copies, and potentially better performance.

## Motivation and additional information

See #1926.
## Author's checklist

* [ ] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [ ] I sticked to C++17 features.
* [ ] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/xml/ConfigParser.cpp [hunk 1]
```diff
@@ -59,14 +59,14 @@ void OnStartElementNs(
     auto        valueEnd   = reinterpret_cast<const char *>(attributes[index + 4]);
     std::string value(valueBegin, valueEnd);
 
-    attributesMap[attributeName] = decodeXML(value);
+    attributesMap[attributeName] = decodeXML(std::move(value));
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:62]: This is a strong indicator that `decodeXML` should take a `string_view` and return a decoded owning `std::string`.

File: src/xml/ConfigParser.cpp [hunk 2]
```diff
@@ -59,14 +59,14 @@ void OnStartElementNs(
...
   }
 
   auto pParser = static_cast<ConfigParser *>(ctx);
 
   std::string sPrefix(prefix == nullptr ? "" : reinterpret_cast<const char *>(prefix));
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:67]: `sPrefix` is either empty or the prefix. This can also be a `string_view`.

File: src/xml/ConfigParser.cpp [hunk 3]
```diff
@@ -59,14 +59,14 @@ void OnStartElementNs(
...
 
-  pParser->OnStartElement(reinterpret_cast<const char *>(localname), sPrefix, attributesMap);
+  pParser->OnStartElement(std::move(reinterpret_cast<const char *>(localname)), std::move(sPrefix), attributesMap);
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:69]: Both can be non-owning, so we change OnStartElement to take two string_views

File: src/xml/ConfigParser.cpp [hunk 4]
```diff
@@ -117,10 +117,10 @@ void OnFatalErrorFunc(void *userData, const char *error, ...)
 
 precice::logging::Logger ConfigParser::_log("xml::XMLParser");
 
-ConfigParser::ConfigParser(const std::string &filePath, const ConfigurationContext &context, std::shared_ptr<precice::xml::XMLTag> pXmlTag)
+ConfigParser::ConfigParser(std::string_view filePath, const ConfigurationContext &context, std::shared_ptr<precice::xml::XMLTag> pXmlTag)
     : m_pXmlTag(std::move(pXmlTag))
 {
-  readXmlFile(filePath);
+  readXmlFile(std::string(filePath));
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:123]: Why does readXmlFile need an owning string in the interface?

Same below

File: src/xml/ConfigParser.cpp [hunk 5]
```diff
@@ -210,7 +210,7 @@ auto gatherCandidates(const std::vector<std::shared_ptr<XMLTag>> &DefTags, std::
 
 void ConfigParser::connectTags(const ConfigurationContext &context, std::vector<std::shared_ptr<XMLTag>> &DefTags, CTagPtrVec &SubTags)
 {
-  std::unordered_set<std::string> usedTags;
+  std::unordered_set<std::string_view> usedTags;
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:213]: This works only under the assumption that the tags don't change. Let's keep this `std::string` to make this more robust to future changes.

File: src/xml/Printer.cpp [hunk 6]
```diff
@@ -317,7 +317,7 @@ std::ostream &printMD(std::ostream &out, const XMLTag &tag, int level, std::map<
         out << "* [" << heading << "](" << link << ") `" << subtag->getOccurrenceString(subtag->getOccurrence()) << "`\n";
       } else {
         auto &tags = groupedTags[ns];
-        tags.emplace_back("[" + subtag->getName() + "](" + link + ") `" + subtag->getOccurrenceString(subtag->getOccurrence()) + "`");
+        tags.emplace_back("[" + subtag->getName() + "](" + link + ") `" + std::string(subtag->getOccurrenceString(subtag->getOccurrence())) + "`");
```
Comment (fsimonis) [inline src/xml/Printer.cpp:320]: Way cleaner alternative would be to using `fmt::format` here.

**Review by fsimonis — COMMENTED** [review]
Hi there,

I did a first iteration on your PR and already learned some things. :+1:

File: src/xml/ConfigParser.cpp (hunk 4 above)
Comment (Durganshu) [reply src/xml/ConfigParser.cpp:123]: Yes, your point is reasonable. 

The reason I was doing an explicit call at the interface was to make it compatible with std::ifstream here. 

https://github.com/precice/precice/blob/a8d2b233c640bdd86f67f036313bed5aeda4038e/src/xml/ConfigParser.cpp#L172

But it is not a clean way of passing it to the interface like this.

File: src/xml/ConfigParser.cpp [hunk 7]
```diff
@@ -169,7 +171,7 @@ int ConfigParser::readXmlFile(std::string const &filePath)
   SAXHandler.error          = OnErrorFunc;
   SAXHandler.fatalError     = OnFatalErrorFunc;
 
-  std::ifstream ifs(filePath);
+  std::ifstream ifs{std::string(filePath).c_str()};
```
Comment (fsimonis) [inline src/xml/ConfigParser.cpp:174]: I don't think this is worth it. Let's change the `filePath` back to `std::string`.

The `ifstream` accepts `std::string`, as it guarantees to be zero-terminated, which the system C-APIs require. `string_view` doesn't fullfill this guarantee.

File: src/xml/Printer.cpp [hunk 8]
```diff
@@ -314,10 +315,16 @@ std::ostream &printMD(std::ostream &out, const XMLTag &tag, int level, std::map<
 
       const auto ns = subtag->getNamespace();
       if (ns.empty()) {
-        out << "* [" << heading << "](" << link << ") `" << subtag->getOccurrenceString(subtag->getOccurrence()) << "`\n";
+        out << fmt::format("[* {}]({}) `{}`",
+                           heading,
+                           link,
+                           std::string(subtag->getOccurrenceString(subtag->getOccurrence())));
```
Comment (fsimonis) [inline src/xml/Printer.cpp:321]: This should work without conversion.
```suggestion
                           subtag->getOccurrenceString(subtag->getOccurrence()));
```

**Review by fsimonis — APPROVED** [review]
This looks ready now. Thanks for the contribution!

**Pull request merged**